pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,231 @@
+	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+type rangeInfo struct {
+	childAddr     uint64
+	parentAddr    uint64
+	addrSpaceSize uint64
+}
+
+type gpioChip struct {
+	name     string
+	dtNode   string
+	physAddr uint64
+	virtAddr uint64
+	chipSize uint32
+	// memfd     uint32
+}
+
+var INVALID_ADDR uint64 = math.MaxUint64
+
+const gpioName = "gpio0"
+const dtBaseNodePath = "/proc/device-tree"
+
+// Sets up GPIO Pin Memory Access by parsing the device tree for relevant address information
+func (b *pinctrlpi5) pinControlSetup() error {
+	nodePath, err := b.findPathFromAlias(gpioName) // this ("gpio") is hardcoded now, we will fix that later!
+	if err != nil {
+		b.logger.Errorf("error getting raspi5 GPIO nodePath")
+	}
+
+	err = b.setGPIONodePhysAddr(nodePath)
+	if err != nil {
+		b.logger.Errorf("error getting raspi5 GPIO physical address")
+	}
+	return err
+}
+
+// We look in the 'aliases' node at the base of proc/device-tree to determine the full file path required to access our GPIO Chip
+func (b *pinctrlpi5) findPathFromAlias(nodeName string) (string, error) {
+
+	dtNodePath := dtBaseNodePath + "/aliases/" + nodeName
+	nodePathBytes, err := os.ReadFile(dtNodePath)
+	if err != nil {
+		return "", fmt.Errorf("Error reading directory: %w", err)
+	}
+
+	// convert readFile output from bytes -> string format
+	nodePath := fmt.Sprintf("%s", nodePathBytes)
+	return nodePath, err
+}
+
+/*
+A child Node will call this method ON ITS PARENT to determine how many cells denote parent address, child address, addr space size when
+reading its ranges or reg properties.
+
+returns:
+
+	numPAddrCells:	number of 32 bit chunks needed to represent the parent address
+	numAddrSpaceCells:	number of 32 bit chunks needed to represent the size of the parent address space
+*/
+func getNumAddrSizeCellsInfo(parentNodePath string) (uint32, uint32, error) {
+
+	// get #address - cells info for child node using the parent Node
+	npaByteContents, err := os.ReadFile(parentNodePath + "/#address-cells")
+	if err != nil {
+		return 0, 0, fmt.Errorf("trouble getting addr cells info for %s: %w\n", parentNodePath, err)
+	}
+	numPAddrCells := binary.BigEndian.Uint32(npaByteContents[:4])
+
+	// get #size - cells info for child node using the parent Node
+	npsByteContents, err := os.ReadFile(parentNodePath + "/#size-cells")
+	if err != nil {
+		return 0, 0, fmt.Errorf("trouble getting size cells info for %s: %w\n", parentNodePath, err)
+	}
+	numAddrSpaceCells := binary.BigEndian.Uint32(npsByteContents[:4]) // reading 4 bytes because the number is represented by 1 uint32. 4bytes * 8bits/byte = 32 bits
+	return numPAddrCells, numAddrSpaceCells, err
+}
+
+// Reads the /reg file and converts the bytestream into a uint64 representing the GPIO Node's physical address within its parent's space. (pre address mapping)
+func getRegAddr(childNodePath string, numPAddrCells uint32) (uint64, error) {
+
+	//newPath := "/proc/device-tree/axi/pcie@120000/rp1/gpio@d0000/reg"
```

Comment:
why is this commented out?

Reply:
I should delete this, it was a hardcoded workaround I was using for testing before I used the reg expression to sanitize the filepath while testing